modules/relevadRtdProvider.js

```diff
@@ -0,0 +1,441 @@
+          }
+        });
+      }
+    } catch (e) {
+      logError(e);
+    }
+  }
+
+  // Set per-bidder RTD
+  const adUnits = reqBids.adUnits;
+  adUnits.forEach(adUnit => {
+    noWhitelists && deepSetValue(adUnit, 'ortb2Imp.ext.data.relevad_rtd', relevadList);
+
+    adUnit.hasOwnProperty('bids') && adUnit.bids.forEach(bid => {
+      let bidderIndex = (moduleConfig.params.hasOwnProperty('bidders') ? findIndex(moduleConfig.params.bidders, function (i) {
+        return i.bidder === bid.bidder;
+      }) : false);
+      const indexFound = !!(typeof bidderIndex == 'number' && bidderIndex >= 0);
+      try {
+        /**
+         * Adds RTD to the bidder and calls a callback for more bidder-specific processing
+         *
+         * @param      {<type>}    bidderName  The bidder name
+         * @param      {Function}  callback    The budder specific callback
+         */
+        const processBidder = function(bidderName, callback) {
+          let wb = isEmpty(wl) || wl[bid.bidder] === true;
+          if (!wb && !isEmpty(wl[bid.bidder])) {
+            wb = true;
+            // match bidder parameters
+            for (const [key, value] of entries(wl[bid.bidder])) {
+              let params = bid?.params || {};
+              wb = wb && (key in params) && params[key] == value;
+            }
+          }
+          if (wb && !isEmpty(relevadList)) {
+            setBidderSiteAndContent(reqBids.ortb2Fragments?.bidder, bid.bidder, relevadData);
+            callback();
+          }
+        };
+
+        if (
+          !biddersParamsExist ||
+            (indexFound &&
+              (!moduleConfig.params.bidders[bidderIndex].hasOwnProperty('adUnitCodes') ||
+                moduleConfig.params.bidders[bidderIndex].adUnitCodes.indexOf(adUnit.code) !== -1
+              )
+            )
+        ) {
+          switch (bid.bidder) {
+            case 'appnexus':
+            case 'appnexusAst':
+            case 'brealtime':
+            case 'emxdigital':
+            case 'pagescience':
+            case 'gourmetads':
+            case 'matomy':
+            case 'featureforward':
+            case 'oftmedia':
+            case 'districtm':
+            case 'adasta':
+            case 'beintoo':
+            case 'gravity':
+            case 'msq_classic':
+            case 'msq_max':
+            case '366_apx':
+              processBidder('appnexus', function() {
+                deepSetValue(bid, 'params.keywords.relevad_rtd', relevadList);
+              });
+              break;
+
+            case 'smart':
+            case 'smartadserver':
+              let target = [].concat(bid.params?.target ? [bid.params.target] : []);
+              processBidder('smart', function() {
+                target = target.concat(relevadList.map(entry => { return 'relevad_rtd=' + entry; }));
+                deepSetValue(bid, 'params.target', target.join(';'));
+              });
+              break;
+
+            case 'ix':
```

Comment:
our goal is to eliminate all this case handling in rtd adapters and move them into bidders. You can pr all the adapters that require special handling of the standard object.

Reply:
Hi,
I took the case handling code almost literally from the already approved RTD provider [sirdataRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/sirdataRtdProvider.js#L230), and my code does exactly the same case handling.
Many other approved RTD providers have very similar case handling code in order to accomodate the non-standard server-side of various bidders, for example:

- [sirdataRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/sirdataRtdProvider.js#L230)
- [permutiveRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/permutiveRtdProvider.js#L253)
- [weboramaRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/weboramaRtdProvider.js#L690)
- [captifyRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/captifyRtdProvider.js#L43)
- [1plusXRtdProvider.js](https://github.com/prebid/Prebid.js/blob/97c149daf215f5f0b178c9fb88f29a3fd9b5b32c/modules/1plusXRtdProvider.js#L246)

I don’t think I have the authority to change other companies  (appnexus, rubicon etc) BidAdapter code.
Simply removing all case handling in my module would cripple it’ functionality vs. the others, and I’d like to keep it competitive.

Considering this, could you please let  the case handling stay in my code?
Please  advise how to proceed

Thank you very much!

Reply:
Addressed your request w.r.t. removing bidder-specific handling' with this commit:
https://github.com/prebid/Prebid.js/pull/9454/commits/6967feb082d477d1f956a505ea8a19d21d8e23b2

Please review it
Thank you